server/util/grpc_client/grpc_client.go

```diff
@@ -216,3 +223,64 @@ func CommonGRPCClientOptions() []grpc.DialOption {
 		}),
 	}
 }
+
+type grpcClientConnPoolCache struct {
+	env         environment.Env
+	connMutex   *sync.RWMutex
+	connPoolMap map[string]interfaces.ClosableClientConn
+}
+
+func RegisterConnPoolCache(env environment.Env) {
+	c := NewGrpcClientConnPoolCache(env)
+	env.SetGrpcClientConnPoolCache(c)
+}
+
+func (cpc *grpcClientConnPoolCache) GetGrpcClientConnPoolForURL(target string) (conn interfaces.ClosableClientConn, shouldClose bool, err error) {
```

Comment:
Having to return and check this shouldClose bool is a bit unfortunate. 

Could you use an LRU to maintain the cache? you can wait for connections to go idle before closing them to avoid closing active connections.

Reply:
maybe i'm missing something, but i didn't use an LRU for now because if a client has just grabbed a connection pool and the LRU decides to evict / close the pool, then the caller will be out of luck--the connection won't get out of an "Idle" state unless the caller tries to do something with it, and otherwise i'm in the same weird space where i need to lock / explicitly close, right?

separately, i didn't see a really awesome way to tell if a handle was held by the caller for a long time or not here.  i guess if i make the assumption that this is only called by bytestream code, then there's a concrete "end" to the connection being used, but i think i kind of still need a real signal that everyone's "done" with connections, right?

Reply:
> maybe i'm missing something, but i didn't use an LRU for now because if a client has just grabbed a connection pool and the LRU decides to evict / close the pool, then the caller will be out of luck--the connection won't get out of an "Idle" state unless the caller tries to do something with it, and otherwise i'm in the same weird space where i need to lock / explicitly close, right?

this is what i was thinking we could do on eviction in background, in pseudo-code:
   check if at least N seconds has passed since the pool was created to make sure client had a chance to use it, if not wait before cleaning up
   wait for pool connections to go idle
   close connections
      
Would that work?